bplustree_test.go

```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
 	t.Logf("\nFinal tree:\n")
 	newtree.Print()
 }
+
+func TestInitStr(t *testing.T) {
+	ctx := Context{lockMgr: nil, maxDegree: maxDegree}
+	var err error
+	// Enable line numbers in logging
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+	newtree, err = NewBplusTree(ctx)
+	if err != nil {
+		t.Errorf("Failed to create new bplustree: %v", err)
+		t.FailNow()
+	}
+}
+
+func InsertRoutineStr(t *testing.T,
+	elem *TestingElemStr,
+	errors []error,
+	index int,
+	wg *sync.WaitGroup) {
+	defer wg.Done()
+	t.Logf("Inserting: %v", elem)
+	err := newtree.Insert(elem)
+	errors[index] = err
+
+	t.Logf("++++++++++++++++++++++++++++++++")
+}
+
+func makeTestingKeyStr(k int) TestingKeyStr {
+	return TestingKeyStr(fmt.Sprintf("key_%d", k))
+}
+
+func TestInsertStr(t *testing.T) {
+	t.Log("Starting insert test..")
+	elems := make([]TestingElemStr, numElems)
+	errors := make([]error, numElems)
+	var wg sync.WaitGroup
+
+	for i := 0; i < numElems; i++ {
+		elem := &elems[i]
+		elem.key = makeTestingKeyStr(i)
+		elem.val = i + 10
+		wg.Add(1)
+		go InsertRoutineStr(t, elem, errors, i, &wg)
+	}
+
+	wg.Wait()
+	for i := 0; i < numElems; i++ {
+		if errors[i] != nil {
+			t.Errorf("Failed to insert: %d: %v", i, errors[i])
+			t.FailNow()
+		}
+	}
+	t.Logf("\nFinal tree:\n")
+	newtree.Print()
+}
+
+func TestSearchStr0(t *testing.T) {
+	t.Log("Starting search tests..")
+
+	// case 0. Look for a key which doesn't exist.
+	testingKey := makeTestingKeyStr(10000)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Exact}
+	result, err := newtree.Search(ss)
+	if err == nil {
+		t.Errorf("Expected to not find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result != nil {
+		t.Errorf("Expected to not get any result but got: %v", result)
+		t.FailNow()
+	}
+
+}
+
+func keyEvaluatorStr(k1 Key, k2 Key) bool {
+	result := k1.Compare(k2)
+	if result < 10 && result > -10 {
```

Comment:
what is the significance of 10 and -10? is this to restrict the evaluations that would yield true? so if you ask for 30 results, you will get maximum of 20?

Reply:
yeah.. if evaluator stops yielding then maxElem doesn't matter.. otherwise maxElems will limit the # of items fetched. User can define the evaluator to their choice.. in our usecase I expect we could have the evaluator which says, give me keys until # of bits difference is less than certain value.. or me 50 closeby keys .. whichever is less.